### PR TITLE
feat: support sql dump

### LINKED ISSUE #2: Support db:structure:dump and db:structure:load
The Postgres DatabaseTasks implementation uses `pg_dump` to dump the database in `db:structure:dump`. CockroachDB doesn't support `pg_dump`, so this plugin should use `DUMP` and its inverse when they're officially supported.

### PR DESCRIPTION
**Pull request by BuonOmo**
Fixes #2

TODO:

- [x] support or explicitly fail for `ActiveRecord.dump_schemas` conf
- [x] explicitly fail having extra_flags
- [x] support for `load_schema` (see #275)
- [x] dump the content of the version table (by reading the code, it is already done and tested in AR, we're fine)

### CODE REVIEW

File: lib/active_record/connection_adapters/cockroachdb/database_tasks.rb [hunk 1]
```diff
@@ -5,7 +5,37 @@ module ConnectionAdapters
...
+            %w(SCHEMAS TYPES).each do |object_kind|
+              ActiveRecord::Base.connection.execute("SHOW CREATE ALL #{object_kind}").each_row { file.puts _1 }
+            end
+
+            ignore_tables = ActiveRecord::SchemaDumper.ignore_tables
+            all_tables = conn.execute("SHOW TABLES").field_values("table_name")
+            tables = all_tables - ignore_tables
+
+            tables.each do
+              create_stmt = conn.execute("SHOW CREATE TABLE #{conn.quote_table_name(_1)}").field_values("create_statement").first
```
Comment (BuonOmo) [inline lib/active_record/connection_adapters/cockroachdb/database_tasks.rb:35]: This implies an ugly amount of back and forth between the cockroach instance and the rails server. I'm looking into a better solution

**Comment by BuonOmo** [comment]
@rafiss it is all yours!

I'm pretty sure there are some edge cases we are not covering as we are basically recreating `pg_dump` in 38 lines (the `structure_dump` method). Which is impossible for sure.

If you already see some of them, I'll implement them, otherwise I think that we should just wait and see issues coming.. I can issue a warning somewhere in the task saying that this behaviour might not dump the whole db if you want

File: lib/active_record/connection_adapters/cockroachdb/database_tasks.rb [hunk 2]
```diff
@@ -5,7 +5,43 @@ module ConnectionAdapters
...
+          end
+
+          case ActiveRecord.dump_schemas
+          when :all, String
+            raise  "Custom schemas are not supported in CockroachDB. " \
+              "See https://github.com/cockroachdb/cockroach/issues/26443."
+          when :schema_search_path
+            if configuration_hash[:schema_search_path]
+              raise  "Custom schemas are not supported in CockroachDB. " \
+                "See https://github.com/cockroachdb/cockroach/issues/26443."
```
Comment (rafiss) [inline lib/active_record/connection_adapters/cockroachdb/database_tasks.rb:20]: custom schemas are well-supported in CockroachDB these days. can we support this case?

File: lib/active_record/connection_adapters/cockroachdb/database_tasks.rb [hunk 3]
```diff
@@ -5,7 +5,43 @@ module ConnectionAdapters
...
+
+          conn = ActiveRecord::Base.connection
+          File.open(filename, "w") do |file|
+            %w(SCHEMAS TYPES).each do |object_kind|
+              ActiveRecord::Base.connection.execute("SHOW CREATE ALL #{object_kind}").each_row { file.puts _1 }
+            end
+
+            ignore_tables = ActiveRecord::SchemaDumper.ignore_tables.to_set
+
+            conn.execute("SHOW CREATE ALL TABLES").each_row do |(sql)|
```
Comment (rafiss) [inline lib/active_record/connection_adapters/cockroachdb/database_tasks.rb:32]: there are a few other statements that we will probably need in order to fully dump everything: `SHOW CREATE ALL TYPES` and `SHOW CREATE ALL SCHEMAS`

**Review by rafiss — CHANGES_REQUESTED** [review]

File: lib/active_record/connection_adapters/cockroachdb/database_tasks.rb (hunk 2 above)
Comment (BuonOmo) [reply lib/active_record/connection_adapters/cockroachdb/database_tasks.rb:20]: There are other places where we don't support it, so I think we should support it all at once, in a dedicated milestone

File: lib/active_record/connection_adapters/cockroachdb/database_tasks.rb (hunk 3 above)
Comment (BuonOmo) [reply lib/active_record/connection_adapters/cockroachdb/database_tasks.rb:32]: We do, see lines 26 27 :)

Comment (rafiss) [reply lib/active_record/connection_adapters/cockroachdb/database_tasks.rb:32]: Thanks for pointing that out :)

File: lib/active_record/connection_adapters/cockroachdb/database_tasks.rb (hunk 2 above)
Comment (rafiss) [reply lib/active_record/connection_adapters/cockroachdb/database_tasks.rb:20]: That sounds good to me!

File: test/cases/tasks/cockroachdb_rake_test.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        returns: ["accounts", "articles"]
+      ) do
+        ActiveRecord::Tasks::DatabaseTasks.structure_dump(config, @filename)
+
+        read = File.read(@filename)
+      end
+
+      read = File.read(@filename)
+      refute read.include?("NOT TODAY"), "The dump file previous content was not overwritten"
+      assert read.include?("CREATE TABLE public.schema_migrations"), "No dump done"
```
Comment (rafiss) [inline test/cases/tasks/cockroachdb_rake_test.rb:41]: should the schema and type statements be tested here too?

**Review by rafiss — APPROVED** [review]
code looks good to me! i just had one testing question.

File: lib/active_record/connection_adapters/cockroachdb/database_tasks.rb (hunk 2 above)
Comment (BuonOmo) [reply lib/active_record/connection_adapters/cockroachdb/database_tasks.rb:20]: #276

File: test/cases/tasks/cockroachdb_rake_test.rb (hunk 4 above)
Comment (BuonOmo) [reply test/cases/tasks/cockroachdb_rake_test.rb:41]: well that is a good question. The tests in the ar test suit are even simpler than that (just checking for a comment). I can add those as they don't cost much to add. But I'm not sure of how secure this is. As stated before, I think an external and complete dump tool would be necessary to ensure some robustness!

Comment (rafiss) [reply test/cases/tasks/cockroachdb_rake_test.rb:52]: very fair point. i think the basic check you've added will at least catch a few regressions, even if it doesn't test each edge case.

**Review by rafiss — APPROVED** [review]
